client/proxy.go

```diff
@@ -0,0 +1,1055 @@
+
+	deleted, _ := proxyConfig.RemoveTcpConnector(proxyName + "-egress-" + address)
+
+	// should this be update or write to config map
+	proxyConfig.UpdateConfigMap(configmap)
+	_, err = cli.KubeClient.CoreV1().ConfigMaps(cli.GetNamespace()).Update(configmap)
+	if err != nil {
+		return fmt.Errorf("Failed to update proxy configmap: %w", err)
+	}
+
+	if isActive(proxyName) && deleted {
+		mc, err := qdr.MarshalRouterConfig(*proxyConfig)
+		if err != nil {
+			return fmt.Errorf("Failed to marshall router config: %w", err)
+		}
+
+		err = ioutil.WriteFile(proxyDir+"/config/qdrouterd.json", []byte(mc), 0644)
+		if err != nil {
+			return fmt.Errorf("Failed to write config file: %w", err)
+		}
+
+		url, err := ioutil.ReadFile(proxyDir + "/config/url.txt")
+		agent, err := qdr.Connect(string(url), nil)
+		if err != nil {
+			return fmt.Errorf("qdr agent error: %w", err)
+		}
+		if err = agent.Delete("org.apache.qpid.dispatch.tcpConnector", proxyName+"-egress-"+address); err != nil {
+			return fmt.Errorf("Error removing tcp connector : %w", err)
+		}
+		agent.Close()
+	}
+
+	return nil
+}
+
+func (cli *VanClient) ProxyExpose(ctx context.Context, options types.ProxyExposeOptions) (string, error) {
+
+	proxyName, err := cli.ProxyInit(ctx, types.ProxyInitOptions{
+		Name:       options.ProxyName,
+		StartProxy: false,
+	})
+	if err != nil {
+		return "", err
+	}
+
+	// Note: expose implicitly creates the service
+	si, err := cli.ServiceInterfaceInspect(ctx, options.Egress.Address)
+	if err != nil {
+		return "", fmt.Errorf("Failed to retrieve service: %w", err)
+	}
+
+	if si == nil {
+		servicePort, err := strconv.Atoi(options.Egress.Port)
+		if err != nil {
+			return "", fmt.Errorf("%s is not a valid port", options.Egress.Port)
+		}
+		err = cli.ServiceInterfaceCreate(context.Background(), &types.ServiceInterface{
+			Address:  options.Egress.Address,
+			Protocol: options.Egress.Protocol,
+			Port:     servicePort,
+		})
+		if err != nil {
+			return "", fmt.Errorf("Unabled to create service: %w", err)
+		}
+	}
+
+	err = cli.ProxyBind(ctx, proxyName, options.Egress)
+	if err != nil {
+		return proxyName, err
+	}
+
+	err = cli.proxyStart(ctx, proxyName)
+	if err != nil {
+		return proxyName, err
+	}
+
+	return proxyName, nil
+}
+
+func (cli *VanClient) ProxyUnexpose(ctx context.Context, proxyName string, address string) error {
+	// Note: no need to unbind as proxy is being deleted
+	// Note: unexpose implicitly removes the cluster service
+	si, err := cli.ServiceInterfaceInspect(ctx, address)
+	if err != nil {
+		return fmt.Errorf("Failed to retrieve service: %w", err)
+	}
+
+	// todo: only remove service if not used, is this necessary
+	if si != nil && len(si.Targets) == 0 && si.Origin == "" {
+		err := cli.ServiceInterfaceRemove(ctx, address)
+		if err != nil {
+			return fmt.Errorf("Failed to removes service: %w", err)
+		}
+	}
+
+	// Note: unexpose will stop and remove proxy independent of bridge configuration
```

Comment:
Another option here would be to only stop the proxy if there are no more service bindings.

Reply:
I had considered that but felt that the spirit of expose/unexpose is to simplify
what commands the user needs to know for the primary use scenario (e.g. expose
process to the skupper network). If a user makes additional bind/forward additions
to the proxy they created via expose then documenting what they will need to
do when they unexpose will get confusing. With the current approach, it will
be clear that if they unexpose, everything goes away. And if they perform
additional bind/forwards then we need to document that they should not
execute the unexpose if they want the proxy config to persist.

Thoughts?

Reply:
I'm fine with that also.